File: Java/src/models/Usuario.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Usuario implements Objeto{
    private int id;
    private String nome;
    private String email;
    private String fone;
    private String senha;
    private EnumUsuarios tipoUsuario;

    public Usuario(int id, String nome, String email, String fone, String senha, EnumUsuarios tipoUsuario){
        setId(id);
        setNome(nome);
        setEmail(email);
        setFone(fone);
        setSenha(senha);
        setTipoUsuario(tipoUsuario);
    }

    public void setId(int id){
        this.id = id;
    }

    public void setNome(String nome){
        if (nome != "") this.nome = nome;
        else throw new IllegalArgumentException("Nome não pode ser vazio");
    }

    public void setEmail(String email){
        if (email == null || email.isEmpty()) {
            throw new IllegalArgumentException("Email inválido");
        }

        //Expressão regular \\ - dispensa o . $ - fim da string ^ - começo da string
        String regex = "^[a-zA-Z0-9._%+-]+@gmail\\.com$";
        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(email);
        
        if (matcher.matches()) {
            this.email = email;
        } else {
            throw new IllegalArgumentException("Email inválido");
        }
    }

    public void setFone(String fone){
        if (fone != "") this.fone = fone;
        else throw new IllegalArgumentException("Numero de telefone não pode ser vazio");
    }

    public void setSenha(String senha){
        if (senha != "") this.senha = senha;
        else throw new IllegalArgumentException("Senha não pode ser vazia.");
    }

    public void setTipoUsuario(EnumUsuarios tipoUsuario){
        this.tipoUsuario = tipoUsuario;
    }

    public int getId(){
        return this.id;
    }

    public String getNome(){
        return this.nome;
    }

    public String getEmail(){
        return this.email;
    }

    public String getFone(){
        return this.fone;
    }

    public String getSenha(){
        return this.senha;
    }

    public EnumUsuarios getTipoUsuario(){
        return this.tipoUsuario;
    }

    public String toString(){
        return String.format("Usuário \n ID = %s \n Nome = %s \n Email = %s \n Fone = %s \n Senha = %s \n Tipo usuário = %s", this.id, this.nome, this.email, this.fone, this.senha, this.tipoUsuario);
    }
}

File: Python/models/Usuario.py
import json
from models.Modelo import Modelo
from enum import Enum
import re

class Usuario:
    def __init__(self, id, nome, email, fone, senha, tipo_usuario):
        self.set_id(id)
        self.set_nome(nome)
        self.set_email(email)
        self.set_fone(fone)
        self.set_senha(senha)
        self.set_tipo_usuario(tipo_usuario)

    def set_id(self, id):
        self.__id = id
    
    def set_nome(self, nome):
        if not nome:
            raise ValueError("Nome vazio")
        self.__nome = nome
    
    def set_email(self, email):
        if not email:
            raise ValueError("Email vazio")
        regex = r'^[a-zA-Z0-9._%+-]+@gmail\.com$'
        if (re.match(regex, email)):
            self.__email = email
        else:
            raise ValueError("Email inválido. Formato válido xxxxxxx@gmail.com")

    def set_fone(self, fone):
        if not fone:
            raise ValueError("Telefone vazio")
        regex = r'^\(\d{2}\) \d{5}-\d{4}$'
        if (re.match(regex, fone)):
            self.__fone = fone
        else:
            raise ValueError("Formato de telefone inválido. Formato válido (xx) xxxxx-xxxx")

    def set_senha(self, senha):
        if not senha:
            raise ValueError("Senha vazia")
        self.__senha = senha

    def set_tipo_usuario(self, tipo_usuario):
        self.__tipo_usuario = tipo_usuario

    def get_id(self):
        return self.__id
    
    def get_nome(self):
        return self.__nome
    
    def get_email(self):
        return self.__email

    def get_fone(self):
        return self.__fone

    def get_senha(self):
        return self.__senha

    def get_tipo_usuario(self):
        return self.__tipo_usuario
    
    def __str__(self):
        return f"ID = {self.get_id()}  |  NOME = {self.get_nome()}  |  EMAIL = {self.get_email()}  |  FONE = {self.get_fone()} | SENHA = {self.get_senha()} | TIPO USUARIO = {self.get_tipo_usuario()}"
    
    def to_dict(self):
        return {
            "id": self.get_id(), "nome": self.get_nome(), "email" : self.get_email(), "fone" : self.get_fone(), "senha" : self.get_senha(), "tipoUsuario" : self.get_tipo_usuario()
        }
    
class Usuarios(Modelo):
    @classmethod
    def salvar(cls):
        with open("../Arquivos/usuarios.json", mode="w") as arquivo:
            json.dump([usuarios.to_dict() for usuarios in cls.objetos], arquivo, indent=6) #vars - converte um objeto em dicionario
            #dump - pega a lista de obejtos e salva no arquivo
            
    @classmethod
    def abrir(cls):
        cls.objetos = []
        try:
            with open("../Arquivos/usuarios.json", mode="r") as arquivo:
                usuarios_json = json.load(arquivo)
                for obj in usuarios_json:
                    u = Usuario(obj["id"], obj["nome"], obj["email"], obj["fone"], obj["senha"], obj["tipoUsuario"])
                    cls.objetos.append(u)    
        except FileNotFoundError:
            pass